openedx/core/djangoapps/config_model_utils/docs/decisions/0001-course-app-config-model.rst

```diff
@@ -0,0 +1,85 @@
+Course App Config Model
+=======================
+
+Status
+------
+
+Proposal
+
+
+Context
+-------
+
+During the development of discussions plugins, a need arose for a system of
+configuration that would allow admins to configure connections to different
+discussion tools, and have course authors select an established configuration
+without needing to share any secrets.
+
+For example, a discussion tool provider might require admins to specify OAuth2
+credentials that will be needed to integrate their tool with the platform. This
+needs to be configured by an admin, and used by a course, but ideally course
+authors should not have access to this information.
+
+It would also be useful to have multiple sets of configuration for different
+discussion tools. A course admin can then select one of the available
+configurations based on their requirements.
+
+Additionally, we might want to prevent certain configurations to be available
+to all courses on a site. So it would be useful to limit configuration options
+to a specific site, org, course, or course run.
+
+This ADR proposes a new abstract configuration model that can be used by course
+apps to store their configuration in a way that can be easily configured by
+admins/staff with different levels of access.
+
+The `StackedConfiguration` model might seem like a close match for this but it
+works in a different way than what is ideal here.
+
+First of all, we do not need stacking in this case, since we are dealing with
+configurations for different discussion tools that may not cleanly stack on top
+of each other.
+
+Second of all, `StackedConfiguration` can only provide one configuration set
+whereas we need support for multiple configuration profiles.
+
+Decision
+--------
+
+We can create a new `CourseAppConfigOptionsModel` Django model that derives
+from `ConfigurationModel`. This model shares some logic with the
+`StackedConfiguration` model but works in a completely different way.
+
+In the `StackedConfiguration` model, values set at different levels, site, org,
+org_course, and course all coalesce into a single configuration. Configuration
+specified at the site level can be overridden by configuration at the org level
+etc. For our use case though, we need to be able to support multiple
+configuration profiles that might be entirely different in shape, and will
+often not be derived from each other.
+
+The new model instead treats each configuration as a separate option available
+to course authors. So an org level configuration will not override a site level
+configuration, but will co-exist with it. Course authors will have both options
+at their disposal when configuring a course.
```

Comment:
> an org level configuration will not override a site level configuration, but will co-exist with it. Course authors will have both options at their disposal when configuring a course.

Is this really necessary though?

You could get much of the same effect using `StackedConfigurationModel` if each plugin declared its own StackedConfigurationModel subclass, and each configuration option was a separate field (defined in `STACKABLE_FIELDS`); though it wouldn't allow a course to pick a site-level configuration of a given plugin instead of an org-level configuration, since the org-level values would always take precedence.

If any provider (like Discourse) is configured at the course/run level, other levels are irrelevant. Obviously I want to use the options configured at the course level, perhaps inheriting an API key from a higher level; certainly I don't need a choice between the course-level config and the org-level config for a given provider, because the course-level config wouldn't exist if it wasn't the one I wanted to use in the course.

So really the only question is: is there ever a time when a provider (say Discourse) is configured at the Site level _and_ the Org level, but a particular course wants to use the site-level version, instead of the Org level version? Maybe, but that seems like a real edge case to me. (More realistic is an org migrating to their own config, and needing old courses to continue to inherit the site-level config)

---------------------------------------

Which brings me to another point: what happens when the site-level or org-level configuration is changed - won't it possibly interrupt many running or archived courses (any that don't specify a course-level config), possibly pointing their discussion board to a new service or different forum on the same service?

---------------------------------------

Does this include any UI / permissions to allow global admins to set global values for the instance, "site admins" to set values for the site, "org admins" to set values for the org, and course instructors to set values for the course?

If not, I'd suggest a much simpler implementation, using regular Django models:

```python
class DiscussionPluginConfiguration:
    """
    Configure a discussion plugin for this Open edX instance, optionally restricted to a site or org.

    Only global staff (sysadmins) can configure these. 

    The "config" can be blank, which means that the plugin is enabled for this Open edX installation/site/org. Multiple configurations may exist for the same plugin, and each course may choose which one to use.
    """
    name = String()  # Name of this configuration e.g. "SomeOrg Discourse"
    plugin = String()  # Which implementation: cs_comments_service, Discourse,  etc.
    config = JsonField()  # API key, endpoint, etc. Nobody other than global staff can view the config data, as it may contain secrets like API keys.
    restrict_to_site = ForeignKey  # Only courses associated with this site(s) may use this config
    restrict_to_org = ForeignKey  # Only courses associated with this organization(s) may use this config

class CourseDiscussionConfiguration:
    """
    Configure a discussion plugin used by a specific course / learning context.
    If an entry is not found for a given course/context, discussions are disabled.
    """
    context_key = LearningContextKey()  # Which course/context this discussion config is for
    enabled = BooleanField()
    global_config = ForeignKey(DiscussionPluginConfiguration)
    course_config = JsonField()  # Overrides like course-specific settings, API key (if different, or if not specified globally), etc.
```

This approach:
* Requires no new abstractions
* Makes minimal use of inheritance
  - site/org is only checked as a permissions/security/validation check when an update to a given course's `CourseDiscussionConfiguration` is made. Inheritance checks are not needed when reading config.
  - This means no complex caching/inheritance logic is needed. The merged configuration for a given course can be queried from a single trivial `CourseDiscussionConfiguration.objects.select_related('global_config').get(context_key=context_key)`
* Has a very clear permissions scheme: only global admins can enable a plugin system-wide or for a particular course/org; they can also (optionally) set default secrets like API key, OAuth creds, etc. but don't have to. Only global admins can see the config values in the `DiscussionPluginConfiguration`, and only course staff can see the values in the `CourseDiscussionConfiguration`.
* Each course always explicitly selects a plugin+configuration to use. If the system-wide configuration for the plugin already sets the required API key etc., then no additional configuration is needed, but the course can always set their own API key or override any other options as needed.
   - This is important to ensure that actions like adding a new org-level configuration don't suddenly inherit down and change what forum is being used for a running course.

I'm not married to this approach, just wanted to suggest it as this PR seems very complex. In any case I'd also suggest starting with a simple configuration paradigm and getting the actual discussion plugins and UI working and implemented before spending a lot more time on configuration helpers and inheritance. The configuration needs may change once more stuff is implemented.

Reply:
> You could get much of the same effect using `StackedConfigurationModel` if each plugin/provider (Discourse, etc.) declared its own StackedConfigurationModel subclass, and each configuration option was a separate field (defined in `STACKABLE_FIELDS`); though it wouldn't allow a course to pick a site-level configuration of a given plugin instead of an org-level configuration, since the org-level values would always take precedence.

StackedConfiguration does too much. It automatically stacks and merges fields across multiples levels and that is not what we want here. Simply taking the highest precedence version is enough. 

> Which brings me to another point: what happens when the site-level or org-level configuration is changed - won't it possibly interrupt many running or archived courses (any that don't specify a course-level config), possibly pointing their discussion board to a new service or different forum on the same service?

This is by design, to allow a changing a config (such as the API key for a service) in one place and have it apply across all services. The plugin that a config is linked to can't be changed, however the URL might. In this case though, that might be desired. If it isn't desired, then we need to warn when performing such an action ("You are modifying a configuration used by X active courses, are you sure you want to continue?"). 

> Does this include any UI / permissions to allow global admins to set global values for the instance, "site admins" to set values for the site, "org admins" to set values for the org, and course instructors to set values for the course?

Not yet. Currently, superadmins with access to this page can set it for all levels. The rest can come as a future enhancement. 

> Each course always explicitly selects a plugin+configuration to use. If the system-wide configuration for the plugin already sets the required API key etc., then no additional configuration is needed, but the course can always set their own API key or override any other options as needed. 

This is already true of the current PR, I've called it a provider instead of plugin, but that sets which plugin this config is tied to, and it cannot be changed. 

The problem with any kind of merging (and the reason why I didn't just use stackedconfig) is that only the plugin can know how the configuration is structured, and can be merged. Perhaps the plugin can provide its own merge function that makes it complicated for an admin to understand how the `course_config` override will work, or the `course_config` can just override entirely (i.e. mask) the `global_config` in which case it might make more sense to make these an either/or option, either pick a pre-existing config, or use a custom config, in which case how do you specify a plugin?

Reply:
Ok great.

My main suggestion was not to use StackedConfiguration, but the two-level approach I suggested, which doesn't require [a new abstraction](https://github.com/edx/edx-platform/blob/24676099d84f5f94e0bcf2d3e1bac0081ba71dd9/openedx/core/djangoapps/config_model_utils/models.py#L390-L394) nor [complex database queries](https://github.com/edx/edx-platform/blob/24676099d84f5f94e0bcf2d3e1bac0081ba71dd9/openedx/core/djangoapps/config_model_utils/models.py#L569-L591), and which allows out-of-the box use of the django admin UI by separating admin-level settings and user-level settings.

> The problem with any kind of merging (and the reason why I didn't just use stackedconfig) is that only the plugin can know how the configuration is structured, and can be merged.

Re merging, I think if you just say that any top-level keys defined in the `CourseDiscussionConfiguration.course_config` will completely overwrite any similar keys in the higher-level `DiscussionPluginConfiguration.config`, if any, then that is workable with any plugin and plugin authors can structure their config format accordingly, and users/admins will understand because it's very clear and works exactly the same way as other configuration keys in the system. (Like `private.py` which overrides `lms.yml` which overrides `common.py`, and in each case the top level keys completely replaces those of previous levels. This is exactly how all django settings work in general, how other [django app plugins](https://github.com/edx/edx-django-utils/tree/master/edx_django_utils/plugins) settings work, and so on.)

Besides, if users override a field like "API key" and forget to also override "endpoint URL", that's a simple error they can fix - it's not something we need to enforce in the configuration system design, by requiring those both to be set at the same time or neither.

However, if what you've got works, and you think the cost of introducing a new abstraction is offset by the benefits and future additional uses, then go with that.

Reply:
@bradenmacdonald I updated this PR to use your suggested approach last sprint.

Reply:
Cool, hope it works well 🤞